src/auto/injector.js

```diff
@@ -843,15 +832,12 @@ function createInjector(modulesToLoad, strictDi) {
         fn = fn[fn.length - 1];
       }
 
-      if (!isClass(fn)) {
+      try {
         // http://jsperf.com/angularjs-invoke-apply-vs-switch
-        // #5388
         return fn.apply(self, args);
-      } else {
-        args.unshift(null);
-        /*jshint -W058 */ // Applying a constructor without immediate parentheses is the point here.
-        return new (Function.prototype.bind.apply(fn, args));
-        /*jshint +W058 */
+      } catch (e) {
```

Comment:
Isn't this quite risky? We're essentially invoking the same code twice, regardless of why this threw an exception. E.g. if the code that's called is `killAllHumansOnSecondCall` (or more realistically, `submitShoppingCartBuyThenThrowException`), wouldn't we accidentally buy two shopping carts?

I think this should at least inspect the thrown exception to make sure class vs no class was the cause.

Also note that `try/catch` brings you onto the non-optimized path in every VM - I'd suspect that the previous solution was better?